src/objective_functions/weight_regularization/l2.cpp

```diff
@@ -30,6 +30,37 @@
 #include "lbann/utils/cublas_wrapper.hpp"
 #endif // LBANN_HAS_CUDNN
 
+namespace {
+
+  /** Compute the entry-wise sum of squares of a local matrix. */
+  EvalType sum_of_squares(const Mat& mat) {
+    const El::Int height = mat.Height();
+    const El::Int width = mat.Width();
+    const El::Int ldim = mat.LDim();
+    const auto& __restrict__ buf = mat.LockedBuffer();
+    EvalType sqsum = EvalType(0);
+    if (ldim == height) {
+      // Parallelize single loop if data is contiguous
+      const El::Int size = height*width;
+      #pragma omp parallel for reduction(+:sqsum)
+      for (El::Int i = 0; i < size; ++i) {
+        const EvalType val = buf[i];
+        sqsum += val * val;
+      }
+    } else {
+      // Parallelize double loop if data is not contiguous
+      #pragma omp parallel for reduction(+:sqsum) collapse(2)
+      for (El::Int j = 0; j < width; ++j) {
+        for (El::Int i = 0; i < height; ++i) {
+          const EvalType val = buf[i + j*ldim];
+          sqsum += val * val;
+        }
+      }
+    }
+    return sqsum;
+  }
+
+} // namespace
```

Comment:
This functionality used to be the private class function `local_squared_l2_norm`. That name is not as clear as `sum_of_squares` and I prefer to hide implementation details from the user.